### ISSUE TITLE
Document journey of first use

### ISSUE THREAD

**Issue by bonndan** [dedica-team/nivio#29]
Review the documentation, what is missing? What can be done to improve the experience of first use?

**Comment by bonndan** [comment]
Please work on the develop branch

**Comment by basseur** [comment]
First use journey using _develop branch_ (Part 1)
========================================
1. Following the documentation I ran `docker run -e DEMO=1 bonndan/nivio:latest`, which did not work as expected.

2. Ran `docker run -e DEMO=1 bonndan/nivio:develop` as a guess, which failed with the following exception:
```
Exception in thread "main" java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:50)
	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:51)
Caused by: de.bonndan.nivio.input.ReadingException: Failed to create an environment from file /src/test/resources/example/example_env.yml
	at de.bonndan.nivio.input.EnvironmentFactory.fromYaml(EnvironmentFactory.java:35)
	at de.bonndan.nivio.input.FileChangeProcessor.process(FileChangeProcessor.java:31)
	at de.bonndan.nivio.Application.main(Application.java:52)
	... 8 more
Caused by: java.nio.file.NoSuchFileException: /src/test/resources/example/example_env.yml
	at java.base/sun.nio.fs.UnixException.translateToIOException(Unknown Source)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(Unknown Source)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(Unknown Source)
	at java.base/sun.nio.fs.UnixFileSystemProvider.newByteChannel(Unknown Source)
	at java.base/java.nio.file.Files.newByteChannel(Unknown Source)
	at java.base/java.nio.file.Files.newByteChannel(Unknown Source)
	at java.base/java.nio.file.Files.readAllBytes(Unknown Source)
	at de.bonndan.nivio.input.EnvironmentFactory.fromYaml(EnvironmentFactory.java:25)
	... 10 more
```

3. Followed the _Development Setup_

    1. `git clone git@github.com:bonndan/nivio.git && cd nivio` worked effortlessly.
    2. `mvn clean package` proceded with a few warnings and errors.
    3. `docker build -t nivio:latest` did not work because of missing argument for a path. `docker build -t nivio:latest .` was succesful (see Issue https://github.com/bonndan/nivio/issues/31).
    4. `docker run -e SEED=/tmp/inout.yml --mount type=bind,source="$(pwd)"/src/test/resources/example,target=/tmp -p 8089:8080 nivio:latest` worked. After the following output
  ```
  2019-10-11 15:23:14.703  INFO 1 --- [ taskExecutor-1] de.bonndan.nivio.input.DirectoryWatcher  : Starting directory watcher on /tmp
  ```
  nothing esle happened and I did not know how to proceed. I tried opening _http://localhost:8080/render/nivio:example/graph.png_, which was not available.

**Comment by bonndan** [comment]
Great, thanks for the examination so far. I've pushed some changes to develop which fix a few things and improves the development setup section in the readme.

**Comment by basseur** [comment]
Nice. Previously, when opening [localhost:8080](http://localhost:8080), there was a failed connection. Now I can see the request being logged in the terminal and the JSON is returned. I found the report, which seems to be rendered correctly.

I am now wondering how I should (and if I even can) try this on _real_ projects. So far I have failed. My user feedback as of now: I am a little frustrated because I don't know how to proceed if I want to analyze another project with nivio. A detailed step-by-step guide would be very helpful but is missing.
Since I wasn't sure what to do, I tried providing different yml-files (like SEED=/path/to/the/yml) from my [Advent of Code project](https://github.com/basseur/AdventOfCode), but was met with an InvocationTargetException:

```
2019-10-11 18:40:52.559  INFO 1 --- [           main] de.bonndan.nivio.input.WatcherFactory    : Created directory watcher for url file:/home/jpgoelz/IdeaProjects/AdventOfCode/deployment-adventofcode.yml
Exception in thread "main" java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:50)
	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:51)
Caused by: de.bonndan.nivio.input.ReadingException: Failed to create an environment from file /home/jpgoelz/IdeaProjects/AdventOfCode/deployment-adventofcode.yml
	at de.bonndan.nivio.input.EnvironmentFactory.fromYaml(EnvironmentFactory.java:35)
	at de.bonndan.nivio.input.WatcherFactory.lambda$getWatchers$0(WatcherFactory.java:55)
	at java.base/java.util.ArrayList.forEach(Unknown Source)
	at de.bonndan.nivio.input.WatcherFactory.getWatchers(WatcherFactory.java:42)
	at de.bonndan.nivio.Application.main(Application.java:46)
	... 8 more
Caused by: java.nio.file.NoSuchFileException: /home/jpgoelz/IdeaProjects/AdventOfCode/deployment-adventofcode.yml
	at java.base/sun.nio.fs.UnixException.translateToIOException(Unknown Source)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(Unknown Source)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(Unknown Source)
	at java.base/sun.nio.fs.UnixFileSystemProvider.newByteChannel(Unknown Source)
	at java.base/java.nio.file.Files.newByteChannel(Unknown Source)
	at java.base/java.nio.file.Files.newByteChannel(Unknown Source)
	at java.base/java.nio.file.Files.readAllBytes(Unknown Source)
	at de.bonndan.nivio.input.EnvironmentFactory.fromYaml(EnvironmentFactory.java:25)
	... 12 more
```

For now I am giving up and await further instructions from you, @bonndan.

**Comment by bonndan** [comment]
If you're running nivio within Docker you need to mount /home/jpgoelz/IdeaProjects/... into the container and provide that path as SEED.

**Comment by basseur** [comment]
**tl;dr:**
I know how to mount the folder and provide a file as SEED. But how do I create the needed yml file correctly? How do I create a yml similar to _inout.yml_ for my own projects?

**long:**
I've been struggling for a while now to understand how to apply _nivio_ to _AdventOfCode_. I think I understand how to mount the corresponding files using `--mount`.

I ran (trying differnet yml files)
`docker run -e SEED=/tmp/service-adventofcode.yml --mount type=bind,source="/home/jpgoelz/IdeaProjects/AdventOfCode,target=/tmp -p 8089:8080 nivio:latest`
(I verified that I understood the parameters correctly by moving and/or renaming _inout.yml_ in my filesystem.)

The features page in the docs says that Nivio collects this information by itself. However, I absolutely don't know how. I am missing a detailed manual for doing this in the docs. After playing around with the contents of _inout.yml_, I believe that I have to write the Application Landscape by hand. Or do I still not understand that correctly? In this case I don't know what to do now. How was _inout.yml_ generated? How do I do this myself for another project? What commands do I need to run or do I need to write my own _inout.yml_?

**Comment by bonndan** [comment]
This is an interesting misunderstanding. What was meant is that Nivio can collect files (or API data) by itself, but cannot (yet?) determine by itself what kind of input is provided. In the case of AdventOfCode, the provided format would be K8s config files, which are not yet supported (only K8s API scan is available as of today).

Inout.yml is proprietary Nivio format and indeed needs to be written by hand. It is meant to enrich data from other sources, because in most cases these sources do not provide enough information. For instance, k8s configs or its API do not say anything about relations between services, like a Java app and it database.

What you can do:

a) add a handwritten landscape yml in order to learn about the underlying model
b) run k8s locally and try to connect Nivio to it (could be difficult, it is alpha and I only tried it with minikube)
c) add a parser for k8s config files (e.g. based on the docker-compose stuff of Nivio), but this is a big task

Please ask question if you're stuck and continue to file issues (great work so far, they help a lot improving the project imho)

**Comment by basseur** [comment]
I went (of course) with (a) and played around a bit. Never mind that this is not a real description of the landscape of AoC. I just wanted to see, if I could add multiple targets and change colors for different groups and stuff like that. I created the file within the AoC-Project and directed source towards the folder. So that worked fine.
One thing I found interesting was that graph.png is generated differently each time. This may have disadvantages. On the other hand, it helped me to get a good version of this landscape. ;)

![graph](https://user-images.githubusercontent.com/32487587/66746249-9b95fa00-ee81-11e9-9ff7-a91366a71fcb.png)

I chose (of course) (a) and played around a little. It doesn't matter that this is not a real description of the AoC landscape. I just wanted to see if I could add multiple targets and change the colors for different groups and stuff. Bidirectional arrows could be helpful.

Getting here wasn't trivial, of course, but we're still at the beginning, so I think it's really cool. The descriptions in _Model and Syntax_ are still a bit thin for a novice user. A GUI tool that could help you with this setup (as an intermediate step to the automatic generation by a parser) would be really nice. With it you could add groups, items, etc. and select the options by dropdown.

**Comment by bonndan** [comment]
Excellent!

Please feel free to improve the docs and post a PR. #24 and #25 are pointing at the GUI idea. In the beginning I was completely against having a GUI, but there is definitely a need for it. So feel free to add your ideas as well.

There graph should not be rendered much differently for the same input (I invested much time into it). If still so, it's not working as desired. Under the hood Nivio is using https://github.com/jgraph/jgraphx, and there is some point where you simply can't get much further without writing your own force-layouted graph algorithm...

**Comment by basseur** [comment]
All right. We should talk about that tonight.

Just for reference, here are a few renders I got by hitting refresh on http://localhost:8080/render/AdventOfCode/graph.png:

![different_renders](https://user-images.githubusercontent.com/32487587/66753716-a9a14600-ee94-11e9-8c69-425f9393c4c4.png)

**Comment by bonndan** [comment]
The random components has been removed several commits earlier in the current develop branch. How to proceed with this ticket?

**Comment by bonndan** [comment]
Closing in favor of #103
